added split methods to polyline

nmaslarinos opened the pull request:
Added `split` and `split_by_length` methods on `compas.geometry.Polyline` to complement the existing divide methods. These methods return a list of Polylines, similar to the split option in Rhino when dividing a curve. Also, updated the docstrings of the divide methods with examples 😊

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

nmaslarinos committed:
added split methods to polyline

nmaslarinos committed:
fix for polyline split methods not returning correct results. Added test functions as well

nmaslarinos committed:
run black

nmaslarinos committed:
removed testing safeguard

nmaslarinos commented:
I have also developed a `break_polyline_at_point` method that opens a polyline on a specified point if closed or splits it in two. I could add it to this pull request since it's all about Polyline methods, or create a new one

tomvanmele commented on src/compas/geometry/primitives/polyline.py:
```diff
@@ -352,14 +352,21 @@ def divide_polyline(self, num_segments):
+        --------
+        >>> polyline = Polyline([(0, 0, 0), (1, 1, 0), (2, 3, 0), (4, 4, 0), (5, 2, 0)])
+        >>> divided_polylines = polyline.divide(3)
+        >>> divided_polyline
```
this variable doesn't seem to be defined

tomvanmele commented on src/compas/geometry/primitives/polyline.py:
```diff
@@ -379,6 +386,18 @@ def divide_polyline_by_length(self, length, strict=True, tol=1e-06):
+        --------
+        >>> polyline = Polyline([(0, 0, 0), (1, 1, 0), (2, 3, 0), (4, 4, 0), (5, 2, 0)])
+        >>> divided_polylines = polyline.divide_by_length(3)
+        >>> divided_polyline
```
same here?

tomvanmele commented on src/compas/geometry/primitives/polyline.py:
```diff
@@ -379,6 +386,18 @@ def divide_polyline_by_length(self, length, strict=True, tol=1e-06):
+
+        >>> polyline = Polyline([(0, 0, 0), (1, 1, 0), (2, 3, 0), (4, 4, 0), (5, 2, 0)])
+        >>> divided_polylines = polyline.divide_by_length(3, strict=False)
+        >>> divided_polyline
```
and here...

nmaslarinos replied:
Sorry about those, I only now realise that for some reason `invoke testdocs` does not perform a check of the files in src, but of the ones in tests

tomvanmele commented on src/compas/geometry/primitives/polyline.py:
```diff
@@ -412,6 +431,98 @@ def divide_polyline_by_length(self, length, strict=True, tol=1e-06):
 
     divide_by_length = divide_polyline_by_length
 
+    def split_polyline_by_length(self, length, strict=True):
```
what is the difference between split and divide?

nmaslarinos replied:
The difference is that divide returns a list of the points that the polyline is divided into. Split returns a list of Polylines. It breaks the polyline into different parts, without disregarding it's geometry.

tomvanmele commented on tests/compas/geometry/test_polyline.py:
```diff
@@ -196,6 +196,140 @@ def test_polyline_split_at_corners(coords, input, expected):
     assert expected == Polyline(coords).split_at_corners(input)
 
 
+@pytest.mark.parametrize(
```
could you add a few boundary cases, and tests that are expected to fail?

tomvanmele commented:
if i understand correctly, the division points are placed based on the length of the original polyline. does this mean the length of the segments of the new polyline is somewhat random?

perhaps this should be clearly stated in the docs, because users might be expecting to get a polyline with segments of the specified length as the result...

tomvanmele commented:
> if i understand correctly, the division points are placed based on the length of the original polyline. does this mean the length of the segments of the new polyline is somewhat random?
> 
> perhaps this should be clearly stated in the docs, because users might be expecting to get a polyline with segments of the specified length as the result...

ha no sorry, i see now that i indeed misunderstood

tomvanmele commented:
LGTM
if you resolve the merge conflicts and fix the docstrings this is good to go...

nmaslarinos committed:
added boundary test cases

nmaslarinos committed:
fixed docstring tests, added checks to input length

nmaslarinos commented:
I added some boundary cases tests in the `test_polyline`, fixed the docstrings which should now work, and added some additional test to the input length (to accept only values larger than zero and smaller than the length of the polyline)

nmaslarinos committed:
Merge branch 'main' into polyline-split-functions

tomvanmele commented:
@nmaslarinos almost there :) your test file has some formatting issues (https://github.com/compas-dev/compas/actions/runs/5297432096/jobs/9589200519?pr=1152). other than that it is good to go...

nmaslarinos committed:
format with black

nmaslarinos commented:
Seems like after formating with black, there is still an [error on the mac-os 3.7 build](https://github.com/compas-dev/compas/actions/runs/5309441021/jobs/9610099589) __FAILED tests/compas/datastructures/test_graph.py::test_graph_networkx_conversion - ModuleNotFoundError: No module named '_bz2'__ . Any idea what this could be?

tomvanmele commented:
no worries this is a github hickup

tomvanmele commented:
i will rerun and the problem probably goes away (had the same a few days ago)

nmaslarinos commented:
I tried to do that, but for the love of me I couldn't find the rerun button 🙄
